src/cryojax/io/_mrc.py

```diff
@@ -3,66 +3,173 @@
+) -> Float[np.ndarray, "..."]:
+    """Read MRC data to a numpy array.
 
     **Arguments:**
 
     'filename' : Path to data.
 
+    `mmap`: Whether or not to open the data as a `numpy.memmap` array.
+
     **Returns:**
 
-    'data' : The array stored in the MRC file.
+    'array' : The array stored in the Mrcfile.
+    """
+    array = _read_array_from_mrc(filename, get_spacing=False, mmap=mmap)
+
+    return array
+
+
+def write_image_stack_to_mrc(
+    image_stack: Float[Array, "M N1 N2"],
+    pixel_size: Float[Array, ""] | Float[np.ndarray, ""] | float,
+    filename: str | Path,
+    overwrite: bool = False,
+    compression: Optional[str] = None,
+):
+    """Write an image stack to an MRC file.
+
+    **Arguments:**
+
+    `image_stack`: The image stack, where the leading dimension indexes the image.
+
+    `pixel_size`: The pixel size of the images in `image_stack`.
+
+    `filename`: The output filename.
+
+    `overwrite`: If `True`, overwrite an existing file.
+
+    `compression`: See `mrcfile.new` for documentation.
+    """
+    # Convert image stack and pixel size to numpy arrays.
+    image_stack = np.asarray(image_stack)
+    pixel_size = np.asarray(pixel_size)
+    # Convert image stack to MRC xyz conventions
+    image_stack = image_stack.transpose(0, 2, 1)
+    # Create new file and write
+    with mrcfile.new(filename, compression=compression, overwrite=overwrite) as mrc:
+        mrc.set_data(image_stack)
+        mrc.set_image_stack()
+        mrc.voxel_size = (1.0, pixel_size, pixel_size)
+
+
+def write_volume_to_mrc(
+    voxel_grid: Float[Array, "N1 N2 N3"],
+    voxel_size: Float[Array, ""] | Float[np.ndarray, ""] | float,
+    filename: str | Path,
+    overwrite: bool = False,
+    compression: Optional[str] = None,
+):
+    """Write a voxel grid to an MRC file.
+
+    **Arguments:**
+
+    `voxel_grid`: The voxel grid as a JAX array.
+
+    `voxel_size`: The voxel size of the `voxel_grid`.
+
+    `filename`: The output filename.
+
+    `overwrite`: If `True`, overwrite an existing file.
+
+    `compression`: See `mrcfile.new` for documentation.
     """
+    # Convert volume and voxel size to numpy arrays.
+    voxel_grid = np.asarray(voxel_grid)
+    voxel_size = np.asarray(voxel_size)
+    # Convert volume to MRC xyz conventions
+    voxel_grid = voxel_grid.transpose(2, 1, 0)
+    # Create new file and write
+    with mrcfile.new(filename, compression=compression, overwrite=overwrite) as mrc:
+        mrc.set_data(voxel_grid)
+        mrc.set_volume()
+        mrc.voxel_size = (voxel_size, voxel_size, voxel_size)
+
+
+def _read_array_from_mrc(
+    filename: str | Path, get_spacing: bool, mmap: bool
+) -> Float[np.ndarray, "..."] | tuple[Float[np.ndarray, "..."], float]:
     # Read MRC
-    with mrcfile.open(filename) as mrc:
-        data = np.asarray(mrc.data, dtype=float)
+    open = mrcfile.mmap if mmap else mrcfile.open
+    with open(filename, mode="r") as mrc:
```

Comment:
I would include options for `permissive=True` - since it's not uncommon with older mrc files in the EMDB that they contain bad header data where mrcfile.open will fail.

See https://mrcfile.readthedocs.io/en/stable/usage_guide.html

Reply:
Cool, this is really good to know. Will add this option. I had to do a merge while working on different branches, but most io updates will be coming from the starfile branch…

The mrc loading I’ve been struggling with a bit because I’ve been using different axis conventions than mrc format… at this point it might be good to switch to z, y, x axis ordering (rather than x, y, z like I’ve been using), but it’s a real pain to convert. Will be working on starfile compatibility in the coming weeks so it should become clear what the right thing to do is.